### PR TITLE
[Pruning] Structured pruning 2x4.

### PR DESCRIPTION
**Pull request by wwwind**
This is the draft of the implementation of structured pruning 2x4:  we set to zero 2 elements in the block of 4 with the smallest absolute magnitude. 

API change:
there is additional parameter: 'sparsity_m_by_n'. When it is set to True, then the layer is pruned using this type of sparsity.
`pruning_params.update({'sparsity_m_by_n': (2, 4)})`

There are several assumptions done in this implementation:

- we apply this type of pruning only to Conv2D and Dense layers;
- in case of Conv2D, pruning is done in input channels dimension, in case of Dense layer - height dimension;
this way we have the required pattern in tflite file.
- if the weight dimension is not divisible by 4, then we do padding;

Added a simple example of CNN on mnist: mnist_e2e_sparsity2x4.py
Added a small tool that checks weights in .tflite file: check_sparsity_m_by_n.py
Example of usage:
`python tensorflow_model_optimization/python/core/sparsity/keras/tools/check_sparsity_m_by_n.py --model_tflite=test_model.tflite --m_by_n=2,4` 

Our preliminary results of testing:
<html><body>
<h4>Experiments with 2:4 sparsity:</h4>

Model | Base | 2:4 sparsity | unstructured sparsity (prunning)
-- | -- | -- | --
DS-CNN-L | **keras model (FP32)**: 95.233% **ds_cnn_l_fp32.tflite**: (accuracy) 95.233%  (size in bytes) 1651768 -> 948513 | **prune stripped keras model** (FP32): 94.538% **ds_cnn_l_fp32.tflite:** (accuracy) 94.538%  (size in bytes) 1652252 -> 983495 |  **prune stripped keras model (FP32)**: 94.640% **ds_cnn_l_fp32.tflite:** (accuracy) 94.640%  (size in bytes) 1652144 -> 949039
Inception-v3 | keras model (FP32): 77.819% | (50% sparsity, 1 epoch, lr: 1e-5) **prune stripped keras model** (FP32) 76.933% | (50% sparsity, 2 epoch, lr: 1e-5) **prune stripped keras model** (FP32) 77.402%
Mobilenet-V2 | keras model (FP32):71.77% | prune stripped keras model (FP32) 64.05% | (50% sparsity)prune stripped keras model (FP32) 67.988%
Mobilenet-V1 | keras model (FP32): 70.976% | pruned keras model (FP32): 67.32% | pruned keras model (FP32): 69.514%
Deit-Tiny | keras model (FP32): 71.6 % | pruned keras model (FP32): 69.19% | -

Testing of 1:2 sparsity on DS-CNN-L model:
<html><body>

DS-CNN-L | unstructured pruning (50% sparsity) | structured pruning (1_by_2 sparsity) 
-- | -- | -- 
Acc. | 94.681% | 94.456% 
File size (bytes) | 1651768 -> 946908 | 1652144 -> 982277 

</body>
</html>

</body>
</html>

### CODE REVIEW

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 from __future__ import print_function
 
 import tensorflow as tf
+import logging
```
Comment (rino20) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:22]: Is it unused?

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py [hunk 2]
```diff
@@ -38,15 +39,23 @@ def __init__(self, training_step_fn, pruning_vars, pruning_schedule,
...
         block. Must be 'AVG' or 'MAX'.
     """
     self._pruning_vars = pruning_vars
     self._pruning_schedule = pruning_schedule
     self._block_size = list(block_size)
     self._block_pooling_type = block_pooling_type
+    self._sparsity_2x4 = sparsity_2x4
     self._validate_block()
 
+    if self._sparsity_2x4:
```
Comment (rino20) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:55]: What about

self._sparsity_2x4 = self._check_if_applicable... if sparsity_2x4 else false

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py [hunk 3]
```diff
@@ -28,7 +29,7 @@ class Pruning(object):
   """Implementation of magnitude-based weight pruning."""
 
   def __init__(self, training_step_fn, pruning_vars, pruning_schedule,
-               block_size, block_pooling_type):
+               block_size, sparsity_2x4, block_pooling_type):
```
Comment (rino20) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:32]: Since this is not scalable when we want to support other types of structured sparsity, we eventually want to generalize this. 

Is there any reason you didn't generalize this to mxn sparsity as you did in pruning_utils.py ?

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py [hunk 4]
```diff
@@ -99,9 +124,31 @@ def _update_mask(self, weights):
...
+    Args:
+      weights: The weight tensor that needs to be masked.
+
+    Returns:
+      new_mask: A numpy array of the same size and shape as weights containing
+        0 or 1 to indicate which of the values in weights should be set to zero.
+        If the returned value is None, the requested mask cannot be created.
+    """
+    window_shape = (1, 4)
+    k = 2
```
Comment (liyunlu0618) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:143]: It would be great if these are configurable from the API level, as Nvidia also proposed 1:2 sparsity in the same [paper](https://arxiv.org/pdf/2104.08378.pdf).

File: tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  converter = tf.lite.TFLiteConverter.from_keras_model(pruned_model)
+  tflite_model = converter.convert()
+
+  tflite_model_path = '/tmp/mnist_2x4.tflite'
+  print('model is saved to {}'.format(tflite_model_path))
+  with open(tflite_model_path, 'wb') as f:
+    f.write(tflite_model)
+
+  print('evaluate pruned model: ')
+  print(keras_test_utils.eval_mnist_tflite(model_content=tflite_model))
```
Comment (liyunlu0618) [inline tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py:147]: Can you add a comment above this line with the final accuracy? Thanks!

File: tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  (x_train, y_train), (
+      x_test,
+      y_test), input_shape = keras_test_utils.get_preprocessed_mnist_data()
+
+  ##############################################################################
+  # Train a model with sparsity 2x4.
+  ##############################################################################
+  pruning_params = {
+      'pruning_schedule':
+          ConstantSparsity(0.5, begin_step=0, frequency=100),
```
Comment (liyunlu0618) [inline tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py:130]: Can you briefly explain the pruning algorithm with this implementation? If I understood correctly, the algorithm Nvidia proposed will only mask the weight once at the beginning of training. The mask is then fixed throughout the training process. Since we have the pruning schedule configured here, does that mean the mask could be updated later?

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py (hunk 3 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:32]: we have not done this as we are interested in this case only and do all testing for it.
but yes, we can do generalized version.

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py (hunk 4 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py]: Hi @liyunlu0618 Thanks for the review. Yes, we can add 1:2 in addition. We will do this.

**Comment by psunn** [comment]
@googlebot I consent.

File: tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py (hunk 5 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py:147]: done

File: tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py (hunk 6 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py:130]: we are doing the same essentially here as Nvidia. we compute mask at the beginning and then it stays the same during the training. i agree that it might be confusing with this scheduler. our intention was to try to increase sparsity gradually, but this work is still under development.

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py (hunk 3 above)
Comment (rino20) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:32]: thanks for the explanation. It looks better to generalize it (make configurable API for other structures) as @liyunlu0618  mentioned below. It would be OK to test mainly on 2:4, since it is the interesting use case for now.

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py [hunk 7]
```diff
@@ -38,13 +38,17 @@ def __init__(self, training_step_fn, pruning_vars, pruning_schedule,
         throughout training.
       block_size: The dimensions (height, weight) for the block sparse pattern
         in rank-2 weight tensors.
+      sparsity_2x4: Boolean that indicates whether we should do sparsity_2x4.
+        We check whether we can do pruning with 2x4 sparsity. If not, then
+        we fallback to the default unstructured pruning for 50%.
```
Comment (rino20) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py:43]: I believe raising error when 2:4 sparsity fails, rather than falling back to unstructured pruning. 
Since user explicitly passed the structured sparsity flag to the API, it is intuitive to fail when it is not possible. Fallback can confuse users who expect structured sparsity but get unstructured one.

**Review by rino20 — COMMENTED** [review]
The overall codes look good, but to make sure the functionality, could you provide testing results on the models popular in production? 

I think 
- one image model (mobilenet, resnet etc,) for code validation for Convolution op, 
- one transformer model (BERT etc.) for validation for dense op. 
would be enough. Is it doable?

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py (hunk 3 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py]: Hi @rino20 I pushed the latest version for the review.

File: tensorflow_model_optimization/python/core/sparsity/keras/test_utils.py [hunk 8]
```diff
@@ -19,7 +19,7 @@
 
 import tensorflow as tf
 
-from tensorflow_model_optimization.python.core.sparsity.keras import prune
+from tensorflow_model_optimization.python.core.sparsity.keras import prune, pruning_utils
```
[the remark this answers is not in the document]
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/test_utils.py]: done

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_utils.py [hunk 9]
```diff
@@ -159,11 +159,48 @@ def factorized_pool(input_tensor,
 
   return tf.squeeze(tf.transpose(width_pooling, perm=[0, 1, 3, 2]))
 
+
+def normalise_tuple(value, name):
```
Comment (rino20) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_utils.py:163]: I am not sure whether this function need to be a shared utility. Ideally, we can  force the input type using typing arguments. 

If you want to use this, please name the function more explicit. From this name, we cannot know it will return a tuple of 2 ints. just suggestion - maybe_convert_to_tuple_of_two_ints

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py (hunk 7 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_impl.py]: Done.

File: tensorflow_model_optimization/python/core/sparsity/keras/prune.py [hunk 10]
```diff
@@ -138,9 +138,11 @@ def prune_low_magnitude(to_prune,
       pruning_policy: (optional) The object that controls to which layers
         `PruneLowMagnitude` wrapper will be applied. This API is experimental
         and is subject to change.
-      sparsity_2x4: (optional) Boolean that indicates whether sparsity 2x4
-        should be applied. In this case, two out of four elements in the
-        weight tensor that have the lowest value are set to zero.
+      sparsity_m_by_n: default None, otherwise a tuple of 2 integers, indicates
+        pruning with m_by_n sparsity, e.g., (2, 4): 2 zeros out of 4 consecutive
+        elements. It check whether we can do pruning with m_by_n sparsity.
+        If not, then it fallback to the unstructured pruning that scheduled with
```
[the remark this answers is not in the document]
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/prune.py]: Done

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_utils.py (hunk 9 above)
Comment (wwwind) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_utils.py]: Done

**Review by rino20 — APPROVED** [review]
Thanks for the work!

**Comment by wwwind** [comment]
Hi @rino20 I resolved the merge conflict and added our numbers for the transformer model Deit-Tiny. 
Please let me know if there is something on our side we could do to get it merged. 
Thanks!

**Review by rino20 — APPROVED** [review]

**Review by daverim — CHANGES_REQUESTED** [review]

**Review by daverim — APPROVED** [review]

**Comment by wwwind** [comment]
Hi @daverim I pushed the requested changes.
Thanks for the review!
